### ISSUE TITLE
Stuttery scrolling, possibly related to KaTeX errors?

### ISSUE THREAD

**Issue by taoeffect** [shd101wyy/markdown-preview-enhanced#85]
I've noticed that while the scrolling in a (very long) document that I'm writing appears smooth in some sections, in others it starts to stutter and radically jump around as I scroll down slowly.

I noticed that it seemed to do this once I started to scroll through this section:

![screen shot 2016-07-25 at 4 23 47 pm](https://cloud.githubusercontent.com/assets/138706/17120905/6466f2c4-5284-11e6-86a0-40185fdc4b02.jpg)

The reason there are KaTeX errors is because I escape certain KaTeX characters, for example: `$\\{-1,0, 1\\}$`

I do this because Wordpress (where this markdown is copied to and then parsed by the JetPack markdown plugin + KaTeX) will treat sequences like `\}` as escape characters, parsing them instead of simply passing them through. Thus, in order to get the backslash to KaTeX for parsing, I have to escape it.

I'd be perfectly happy if you know of a solution to the Wordpress problem, or a solution to the scrolling problem that appears to be caused when KaTeX produces errors (or perhaps, even, for a "Wordpress compatibility" checkbox to be introduced to handle the double-backslash).

Thanks again for your help and for the best Markdown Preview plugin out there!!!! 😄 

```markdown
### The Setup

To avoid confusion, it's necessary to explain _the context_ in which May's Theorem applies, as well as the meaning of some of the terms we'll use. If you notice something missing, please let us know in the comments.[^7]

[^7]: Because this post [is not a PDF](https://twitter.com/taoeffect/status/745741053844561920), and is not behind a paywall, such a thing is possible.

- **Group decision function:** May defines this as a function that takes as input a set of votes and outputs one of: $\\{-1,0, 1\\}$
- **Threshold** or **group threshold:** The amount of votes _in favor_ of a motion as a percentage of the group size, _above which_ the motion will pass. I.e. in majority rule that threshold is $\small50\%$.
- **Voting rule:** For convenience in this blog post only, we use this to refer to a group decision function where all votes carry _equal weight._
- **Majority rule:** A voting rule with a threshold of $\small50\%$.
- **Supermajority rule:** Any voting rule with a threshold $\footnotesize{\gt}\small50\%$ ([typically 2/3 or more](https://en.wikipedia.org/w/index.php?title=Supermajority&oldid=723462259#Common_supermajorities)).
- **Submajority rule:** Any voting rule with a threshold of $\footnotesize{\lt}\small50\%$ . Although the rarest of the rules, some argue there are [good reasons](http://papers.ssrn.com/sol3/papers.cfm?abstract_id=495569) to use submajority rule,[^8] and you might notice that [plurality rules](https://en.wikipedia.org/wiki/Plurality_(voting)) can be a form of submajority rule.
- **Turnout:** The number or percentage of voters who made an actual vote on a motion (not including "abstain" votes).
- **Turnout threshold:** Like threshold, but applied to the turnout only.

[^8]: Note that using a turnout threshold doesn't make as much sense in the context of submajority rule.

May maps English meaning to the output of the group decision function like so:

> We assume $n$ individuals and two alternatives $x$ and $y$. Symbolizing "the $i$th individual prefers $x$ to $y$" by $xP_iy$ and "the $i$th individual is indifferent to $x$ and $y$" by $xI_iy$, we assume that for each $i$ one and only one of the following holds: $yP_ix$, $yI_ix$, or $xP_iy$. With each individual we associate a variable $D_i$ that takes the values $-1, 0, 1$ respectively for each of these situations. Similarly, for the group, we write $D = -1, 0, 1$ according as $yPx$, $yIx$, or $xPy$, i.e., according as the group decision is in favor of $y$, indifference, or in favor of $x$.

In other words:<!-- $-1 =$ "against", $0 =$ "indifferent or indifference", $-1 =$ "in favor". -->

$$\\begin{aligned}-1 &= \\text{against} \\\\ 0 &= \\text{indifferent, indifference} \\\\ 1 &= \\text{in favor}\\end{aligned}$$
```

**Comment by shd101wyy** [comment]
Hi @taoeffect , sorry for my late reply.

I just had a look at jetpack markdown module that you talked about, it seems that there is a bug about backslash.
https://github.com/Automattic/jetpack/issues/3914 .    
I think you need to wait for jetpack team to fix it.  

As for the lagging scrolling, I couldn't figure out why it happens right now.
Try make your file shorter maybe?
I will notice you once I find the solution.

Sorry for the trouble :(

**Comment by taoeffect** [comment]
> I just had a look at jetpack markdown module that you talked about, it seems that there is a bug about backslash.
Automattic/jetpack#3914 .

That particular problem seems possibly related but is different (i.e. mine is not related to revision restoration). Thank you for the link however, I did [write a comment there](https://github.com/Automattic/jetpack/issues/3914#issuecomment-237130545) to make a note about it.

> As for the lagging scrolling, I couldn't figure out why it happens right now.
Try make your file shorter maybe?

I don't think think I can do that? I mean, I have to write up a blog post, and I think a ~326 line post is reasonable?

> I will notice you once I find the solution.

Thank you!

**Comment by shd101wyy** [comment]
Hi, 

Is it convenient for you to send me your markdown file? So that I can locate the cause of the scrolling lag more easily.

As for the KaTeX problem, I have never used jetpack before. So maybe you can try the MathJax plugin (I don't know if there is any or not) instead to see if the same problem still happens.

Thanks

**Comment by taoeffect** [comment]
@shd101wyy Hi! Sorry I accidentally filtered github email notifications and therefore didn't notice that you'd replied.

Certainly I can send you the markdown file, but is there a way I can send it to you privately?

Also, I just had this idea: would it be possible for you to use linear interpolation to make the right side scroll smoothly instead of jumpy-scrolling?

**Comment by shd101wyy** [comment]
My email address is `shd101wyy@gmail.com`.   
Yea I think linear interpolation is a good idea.   
I am quite busy recently as school just started, so I will try to solve this problem when I am free.  

Thanks

**Comment by taoeffect** [comment]
@shd101wyy Thanks! Sent!

**Comment by shd101wyy** [comment]
@taoeffect I have just published a new version(`0.8.3`) that supports smooth scrolling. You can check if that is what you want.  

Cheers!

**Comment by taoeffect** [comment]
@shd101wyy Awesome, it seems to do the smooth scroll when it can but it still stutters on that document that I sent you. And by "stutters" I mean it will literally jump to the bottom of the document and then back up and jerk around all over the place. It seems to do this in all sections of the document now...

**Comment by shd101wyy** [comment]
@taoeffect below is a gif that demos scrolling  
![issue](https://cloud.githubusercontent.com/assets/1908863/18024522/31a6f34a-6bd2-11e6-9396-53d10a21e732.gif)  
I also tried the file that you sent me, and I didn't meet `stutters` that you mentioned.  
Could you please send me a video or gif of the `stutters` situation like you mentioned, I really hope I could solve this issue :)    

Sorry for the trouble.

**Comment by taoeffect** [comment]
@shd101wyy You should have an email from me that shows the stuttery/jumpy scrolling behavior.

**Comment by taoeffect** [comment]
Does it help to know that I have a retina display that looks like 1920x1200 but is, by virtue of being a retina display, actually 3840x2400 (I think)? Also, I edit while having the document basically fullscreen (expanded all the way).

**Comment by shd101wyy** [comment]
Did you click mouse left button after you done scrolling?

**Comment by taoeffect** [comment]
> Did you click mouse left button after you done scrolling?

@shd101wyy I don't know what you mean? What does the mouse button have to do with scrolling? I'm just using two-finger scrolling (same as a scroll wheel).

**Comment by shd101wyy** [comment]
When you click mouse left button at a line, the preview will scroll to the position so that the line in preview and the line where your mouse pointer is at will be matched horizontally. 

(Well, I don't think that is the reason that caused stutters).    

Does stutters happen in all your markdown files or only the one that you sent me?

**Comment by taoeffect** [comment]
> When you click mouse left button at a line, the preview will scroll to the position so that the line in preview and the line where your mouse pointer is at will be matched horizontally. 

When I click on a line it does move where it's scrolled to a little bit, yes.

> Does stutters happen in all your markdown files or only the one that you sent me? 

Seems to mostly happen on that one. I just opened another long one which also has a lot of KaTeX, but none of it is KaTeX that is adjusted for working on Wordpress, and therefore it doesn't show any of the red KaTeX errors. That other file seems to scroll just fine.

Also, shortening the document that I sent you seems to help a bit, but I'm not sure why. I think it might have something to do with the number of images, the KaTeX warnings, or maybe some combination. I really don't know. It's confusing why it works in some documents better than others.

**Comment by taoeffect** [comment]
Woah, holy crap. With the recent update to Atom 1.10.0 the scrolling behavior appear to be suddenly working! :-O

I need to confirm this for sure but if this continues to work I'll be able to close this issue! 😄

**Comment by shd101wyy** [comment]
😂 dont know what happened, but it's a good news 😂

**Comment by taoeffect** [comment]
OK, several days now and it still seems to be working great, and the smooth-scrolling implementation is great! You did a fantastic job on it! 😄 

Happy I can close this issue finally, whatever the reason is for it. :P
